src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -947,4 +983,22 @@ public String concreteIndex() {
             return concreteIndex;
         }
     }
+
+    static class IndexShardReference {
```

Comment:
I made it a reference because we need to remember if we incremented in the PrimaryPhase (see https://github.com/elastic/elasticsearch/pull/10610#discussion_r29253762). I don't understand how making IndexShardReference implement Closeable would help here. Can you elaborate? 
In addition, if I I would override a method like newShardReference that only returns a Closeable than effectively the tests would test if TransportShardReplication work OK with whatever Closeable my test returns, not what IndexShard uses and that seems fishy to me. 
Or maybe I don't get it.